### DISCUSSION TITLE
How to access the native GlobalCoordinateSystem pointer with OpenXR

### DISCUSSION THREAD

**Discussion by Audenrith** [microsoft/MixedRealityToolkit-Unity#10267] [Q&A]
Before Unity 2019.3, one has been able to access the native pointer to the world `SpatialCoordinateSystem` via `UnityEngine.XR.WSA.WorldManager.GetNativeISpatialCoordinateSystemPtr()`.
Switching to Unity 2019 one is able to access the same pointer via `UnityEngine.XR.WindowsMR.WindowsMREnvironment.OriginSpatialCoordinateSystem` using the `Windows Mixed Reality` XR Plugin. 
Can you describe the way of accessing the **native** `SpatialCoordinateSystem` using OpenXR? 

There are some issues which describe how to access a `SpatialCoordinateSystem` as a Unity/C# Object (e.g. [in the Unity forum ](https://forum.unity.com/threads/unityengine-xr-wsa-worldmanager-is-obsolete-what-is-the-alternative.1062509) or [in this repository](https://github.com/microsoft/MixedRealityToolkit-Unity/issues/10082#issuecomment-880980827)) with OpenXR, but the solutions do not describe how to get the **native** pointer (`IntPtr`) (Allocating a `GCHandle` from the C# object obviously can not be used as a `Windows::Perception::Spatial::ISpatialCoordinateSystem^` in c++). 

Is there a way on accessing the native pointer of the origin `SpatialCoordinateSystem` ?

In c++ we are currently casting the pointer via: 
```c++
reinterpret_cast<Windows::Perception::Spatial::ISpatialCoordinateSystem ^>(unitySpatialCoordinateSystemIntPtr);
```
and use it for getting the transform to a specific frame:
```c++
unitySpatialCoordinateSystem->TryGetTransformTo(frameCoordinateSystem);
```

**Comment by Audenrith** [comment]
According to [this comment](https://githubmemory.com/repo/microsoft/MixedRealityToolkit-Unity/issues/10082) one can access the "sceneCoordinateSystem" via: 

```csharp
using namespace Microsoft.MixedReality.OpenXR;
SpatialCoordinateSystem origin = PerceptionInterop.GetSceneCoordinateSystem(UnityEngine.Pose.identity) as SpatialCoordinateSystem;
```

But this does not answer the question on how to get the underlying native pointer. I tried several approaches which all resulted in some error:

The following errors occur:
```csharp
var handle = GCHandle.Alloc(origin , GCHandleType.Pinned); // <-- Error
IntPtr coordinateSystemNative = (IntPtr)(handle);
```
> ArgumentException: Object contains non-primitive or non-blittable data.

```csharp
var handle = GCHandle.Alloc(origin);
IntPtr coordinateSystemNative = (IntPtr)(handle);
// Error in assigning the variable in c++
```

> Access violation reading location
> 
> `__pUnknown->__abi_AddRef();` in `__abi_winrt_ptr_assign(void** __ppTargetArg, const volatile ::Platform::Object^ __objArg)`

```csharp
var handle = GCHandle.Alloc(origin);
IntPtr coordinateSystemNative = handle.AddrOfPinnedObject(); // <-- Error
```

> InvalidOperationException: Handle is not pinned.


```csharp
IntPtr coordinateSystemNative = Marshal.GetIUnknownForObject(origin);
// Error in using the variable in c++
// sceneCoordinateSystem->TryGetTransformTo(frameCoordinateSystem);
```

> Access violation executing location.

Can you tell me, how is the correct way to get this `object` to a `Windows::Perception::Spatial::ISpatialCoordinateSystem^` in `c++`?

**Comment by keveleigh** [comment]
The result from `PerceptionInterop.GetSceneCoordinateSystem(UnityEngine.Pose.identity)` is an `object` that was marshalled from an `IUnknown`. If you're mainly interested in passing that object back to C++, you should be able to pass it back in as-is.

You'll want to use `[MarshalAs(UnmanagedType.IUnknown)] object origin` as the C# interop parameter, and then receive it in C++ as an `IUnknown*` and you should be able to [QI it to `ISpatialCoordinateSystem`](https://docs.microsoft.com/en-us/windows/win32/api/unknwn/nf-unknwn-iunknown-queryinterface(q)) from there.

**Reply by Audenrith** [reply]
Thank you for the clarification.

I managed to get the `object` to C++ and statically cast it to an `IUnknown*` (we have to go through a C interface) but since the `ISpatialCoordinateSystem` requires a `ref` pointer (`an ordinary pointer to a C++/CX mapping ref class or interface class is not allowed`), I didn't manage to perform the `QueryInterface`.

```c++
IUnknown* unityObject;
Windows::Perception::Spatial::ISpatialCoordinateSystem^ coordinateSystem;
unityObject->QueryInterface(&coordinateSystem); // Error: no instance of overloaded function matches the argument list
unityObject->Release();
```

What do I miss here?

**Reply by Audenrith** [reply]
We didn't managed to correctly receive the pointer in c++. After trying different approaches it seems that the pointer we are receiving will reduce the reference count and will be deleted before we can use it. I should mention, that we require to store the pointer permanently (as long as the program is running in Unity) and use it in a separate thread. 

Additionally, we aren't able to use c++ 17 and clr so using a native `GCHandle` object in c++ is probably also not possible.
